_posts/2022-06-17-nativejdb-debugger-for-native-images.adoc

```diff
@@ -0,0 +1,143 @@
+
+
+{empty} +
+
+
+image::nativejdbarch_2.png[alt=NativeJDBArchitecture_2]
+
+
+{empty} +
+
+
+NativeJDB has been using a scaffolding JVM to help it obtain certain static information and speed up development. So currently NativeJDB starts up the program as a Java process as well, in addition to running the native executable. It attaches to that process and suspends it to obtain general static information about the program. In the future, we will remove this scaffolding and replace with information coming from the GDB, though it has been very helpful in rapid prototyping.
+
+
+[#nativejdbinaction]
+== NativeJDB in action!
+
+
+To get started, you'll need the following repos and follow the instructions in each:
+
+
+* https://github.com/nativejdb/nativejdb
+* https://github.com/nativejdb/nativejdbExamples
+
+
+Check out the following video to see NativeJDB in action!
+
+
+{empty} +
+
+
+video::LhTR_ECSaAo[youtube]
+
+
+{empty} +
+
+
+And this video showing NativeJDB running on the GettingStarted Quarkus application:
+
+
+{empty} +
+
+
+video::_9ejxCtRAdg[youtube]
+
+
+[#features]
+== Features
+
+* Ability to show Java source in the editor and step through code
+* Works with https://www.jrebel.com/blog/best-java-ide[IntelliJ] and Java11
+* Works with https://www.graalvm.org/[GraalVM’s] natively compiled images
+* Works with https://github.com/qbicc/qbicc[Qbicc’s] natively compiled images (work in progress)
+* Debugging features using the IDE’s Debug Console itself:
+
+** Suspend / Resume
+** Set Breakpoints (Insert/Enable)
+** Clear Breakpoints (Delete/Disable)
+** Step Over / Step Into / Step Return (work in progress)
+** Stack Frames information in IDE debugger pane
+** Variables (Local + Static) values (work in progress)
+** View of assembly code within a stack frame (work in progress)
+** Multi-threading & thread info
+
+
+NativeJDB does not support hot code replace. Also, very short running programs currently need a Thread.sleep(). This is due to the fact that NativeJDB today uses a scaffolding VM, and needs a little time to attach and suspend it. This problem will disappear when we get rid of the scaffolding in the future. There is also a known issue of breakpoints working only once in loops under some conditions (related to https://github.com/oracle/graal/issues/4379[this] graalvm issue), and the step operation sometimes continues instead.
+
+
+[#conclusion]
+== Conclusion
+
+
+Through this blog, we have showcased a new debugging tool called NativeJDB that allows users to remotely attach and debug a natively compiled Java code.
+
+
+We look forward to getting feedback from you!
```

Comment:
I would add a bit more text here explaining it is a working prototype that enables use of GDB in existing IDE tooling and that we are looking for feedback/suggestions. Maybe even mention idea of Microsoft debug protocol as future potential target ?

Reply:
Sounds good for adding more detail.

The Microsoft Debug Adapter protocol would be useful if we were extending existing IDEs, right? We are at a different point in the design space, where we want to allow existing IDEs to do remote debugging and attach to NativeJDB. So I am inclined not to include that. What do you think?

Reply:
sure, works for me